display_list/display_list_path_effect.h

```diff
@@ -0,0 +1,173 @@
+
+// An enumerated type for the recognized PathEffect operations.
+// In current Flutter we only use the DashPathEffect.
+// And another PathEffect outside of the recognized types is needed
+// then a |kUnknown| type that simply defers to an SkPathEffect is
+// provided as a fallback.
+enum class DlPathEffectType {
+  kDash,
+  kUnknown,
+};
+
+class DlPathEffect
+    : public DlAttribute<DlPathEffect, SkPathEffect, DlPathEffectType> {
+ public:
+  static std::shared_ptr<DlPathEffect> From(SkPathEffect* sk_path_effect);
+
+  static std::shared_ptr<DlPathEffect> From(
+      sk_sp<SkPathEffect> sk_path_effect) {
+    return From(sk_path_effect.get());
+  }
+
+  virtual const DlDashPathEffect* asDash() const { return nullptr; }
+
+ protected:
+  DlPathEffect() = default;
+
+ private:
+  FML_DISALLOW_COPY_ASSIGN_AND_MOVE(DlPathEffect);
+};
+
+/// The DashPathEffect which specifies modifying the path effect, and it
+/// only affects storked paths.
+/// intervals: array containing an even number of entries (>=2), with
+/// the even indices specifying the length of "on" intervals, and the odd
+/// indices specifying the length of "off" intervals. This array will be
+/// copied in Make, and can be disposed of freely after.
+/// count: number of elements in the intervals array
+/// phase: offset into the intervals array (mod the sum of all of the
+/// intervals).
+///
+/// For example: if intervals[] = {10, 20}, count = 2, and phase = 25,
+/// this will set up a dashed path like so:
+/// 5 pixels off
+/// 10 pixels on
+/// 20 pixels off
+/// 10 pixels on
+/// 20 pixels off
+/// ...
+/// A phase of -5, 25, 55, 85, etc. would all result in the same path,
+/// because the sum of all the intervals is 30.
+///
+class DlDashPathEffect final : public DlPathEffect {
+ public:
+  static std::shared_ptr<DlPathEffect> Make(const SkScalar intervals[],
+                                            int count,
+                                            SkScalar phase);
+
+  DlPathEffectType type() const override { return DlPathEffectType::kDash; }
+  size_t size() const override {
+    return sizeof(*this) + sizeof(SkScalar) * count_;
+  }
+
+  std::shared_ptr<DlPathEffect> shared() const override {
+    return Make(pods(), count_, phase_);
+  }
+
+  const DlDashPathEffect* asDash() const override { return this; }
+
+  sk_sp<SkPathEffect> skia_object() const override {
+    return SkDashPathEffect::Make(pods(), count_, phase_);
+  }
+
+  SkScalar* intervals() { return reinterpret_cast<SkScalar*>(this + 1); }
+
+ protected:
+  bool equals_(DlPathEffect const& other) const override {
+    FML_DCHECK(other.type() == DlPathEffectType::kDash);
+    auto that = static_cast<DlDashPathEffect const*>(&other);
+    return count_ == that->count_ && base_equals_(that) &&
+           phase_ == that->phase_;
+  }
+
+ private:
+  const SkScalar* pods() const {
+    return reinterpret_cast<const SkScalar*>(this + 1);
+  }
+
+  bool base_equals_(DlDashPathEffect const* other) const {
```

Comment:
The `base_equals_` methods in the linear gradient classes were shared amongst multiple types of gradients. Since there is only 1 DashPathEffect, just move this code in to the regular `equals_` method or rename it to something more appropriate to its function ("base" doesn't make sense here).

Reply:
This method is unused. Remove it.